### PR TITLE
Add a bulk publish option to the `#publish_topic` API

### PR DESCRIPTION
**Pull request by agrare**
Kafka allows for far more efficient publication of multiple messages by producing many messages then waiting on them after publication.

### CODE REVIEW

**Comment by agrare** [comment]
I don't love the name `publish_topic_multi` because it isn't consistent with the "queue" side which has `publish_message` and `publish_messages`.

I think the issue is in the name `publish_topic` vs `publish_message`.  The first indicates what you're publishing to and the later what type you're publishing.

Ideally I think:
`publish_message_queue` and `publish_messages_queue`
`publish_message_topic` and `publish_messages_topic`

would be more consistent but not sure if we want to tackle that in this PR or not.

**Comment by Fryguy** [comment]
@agrare I agree.  Alternatively, could we change the interface of public_topic to take either a single message or multiple messages?  Then, the impls could handle it under the covers?

**Comment by Fryguy** [comment]
I'm thinking something like:

```ruby
      def publish_topic(*messages)
        message.each { |m| assert_options(m, [:event, :service]) }

        publish_topic_impl(*messages)
      end

# kafka

        def publish_topic_impl(*messages)
          multi = messages.size > 1
          handles = messages.map { |message| raw_publish(multi, *topic_for_publish(message)) }
          handles.each(&:wait)
        end

# stomp

        # Rename publish_topic_impl to publish_topic_single, then...

        def publish_topic_impl(*messages)
          messages.each { |message| publish_topic_single(message) }
        end
```

**Comment by agrare** [comment]
> could we change the interface of public_topic to take either a single message or multiple messages

Ha yes, this is actually how I started out before I realized that the queue side had different methods.  The `assert_options` is a little tricky (it checks that you have the proper keys set in the passed hash).  The `publish_messages` method skips this entirely, the other option would be to `assert_options` on every element of the passed array which is what I would have expected.

**Comment by Fryguy** [comment]
I'm fine with the queue side being different - maybe we can tweak that in a major version.  I feel like having the interface _only_ know about many messages feels more intuitive to use...if you happen to pass 1 we treat it as an array of 1

File: lib/manageiq/messaging/kafka/topic.rb [hunk 1]
```diff
@@ -12,6 +12,11 @@ def publish_topic_impl(options)
           raw_publish(true, *topic_for_publish(options))
         end
 
+        def publish_topic_multi_impl(messages)
+          handles = messages.map { |message| raw_publish(false, *topic_for_publish(message)) }
```
Comment (Fryguy) [inline lib/manageiq/messaging/kafka/topic.rb:16]: I don't know what this first param means, so I'm assuming it means "not-multi"?

Comment (agrare) [reply lib/manageiq/messaging/kafka/topic.rb:16]: The first param means "wait", so for multi you don't want to wait on each message.  Honestly I think we can rip this out of raw_publish and let the _impl methods do the right thing depending on how they're calling it.

File: lib/manageiq/messaging/client.rb [hunk 2]
```diff
@@ -170,9 +170,10 @@ def subscribe_background_job(options)
       # Other options are underlying messaging system specific.
       #
       def publish_topic(options)
-        assert_options(options, [:event, :service])
+        messages = Array.wrap(options)
```
Comment (agrare) [inline lib/manageiq/messaging/client.rb:173]: Had to use `Array.wrap()` here from AS because `Array({:service => "foo"})` was turning into `[[:service, "foo"]]` not `[{:service => "foo"}]`

Comment (agrare) [reply lib/manageiq/messaging/client.rb:173]: Also `publish_topic(*messages)` wasn't handling `publish_topic(inventory_collection.map { |obj| {:whatever} })` aka an actual array passed in so this seemed to handle single hash or an array of hashes better than requiring the caller to splat the args out.  If you think this is too much overhead for the single hash case we could `kind_of?(Array) ? :`

File: lib/manageiq/messaging/kafka/common.rb [hunk 3]
```diff
@@ -21,12 +21,9 @@ def consumer(beginning, options)
           @consumer = kafka_client.consumer
         end
 
-        def raw_publish(wait, body, options)
```
Comment (agrare) [inline lib/manageiq/messaging/kafka/common.rb:24]: I didn't like how `raw_publish(true, ...` and `raw_publish(false, ...` looked as the caller, it wasn't clear what the first arg was doing.  This refactoring is in its own commit if we want to split it out into a separate PR if this is too much change.

File: lib/manageiq/messaging/kafka/common.rb [hunk 4]
```diff
@@ -21,12 +21,9 @@ def consumer(beginning, options)
...
+        def raw_publish(body, options)
           options[:payload] = encode_body(options[:headers], body)
-          producer.produce(options).tap do |handle|
-            handle.wait if wait
-            logger.info("Published to topic(#{options[:topic]}), msg(#{payload_log(body.inspect)})")
```
Comment (agrare) [inline lib/manageiq/messaging/kafka/common.rb:28]: This much logging seemed like overkill (especially to `.info`), we already set `logger` as the `RdKafka` logger.

File: lib/manageiq/messaging/client.rb [hunk 5]
```diff
@@ -170,9 +170,10 @@ def subscribe_background_job(options)
       # Other options are underlying messaging system specific.
       #
```
Comment (Fryguy) [inline lib/manageiq/messaging/client.rb:171]: We might want some documentation here that the incoming `options` are single or Array.

On that note, can we rename `options` to `messages`?  It's a weird name as is.

**Comment by Fryguy** [comment]
A couple more minor questions, but overall LGTM :+1:

File: lib/manageiq/messaging/client.rb (hunk 5 above)
Comment (agrare) [reply lib/manageiq/messaging/client.rb:172]: :+1: yeah good call, changed to messages

File: spec/manageiq/messaging/client_spec.rb [hunk 6]
```diff
@@ -76,6 +76,11 @@
       expect(subject).to receive(:publish_topic_impl)
       subject.publish_topic(:service => 'a', :event => 'e')
     end
+
+    it 'sends an array of messages to a topic' do
+      expect(subject).to receive(:publish_topic_impl).with(kind_of(Array))
```
[the remark this answers is not in the document]
Comment (agrare) [reply spec/manageiq/messaging/client_spec.rb]: :+1: done

File: lib/manageiq/messaging/kafka/common.rb (hunk 3 above)
Comment (kbrock) [reply lib/manageiq/messaging/kafka/common.rb:24]: I like removing the first argument but it feels like a good concept. Can we introduce 2 different methods? Something like `raw_publish_sync` and `raw_publish_blocking`, or call one `raw_publish` and the other with the longer name?

EDIT: if `raw_publish` is an internal method and you moved the `wait` into the caller, then this is very nice.

File: lib/manageiq/messaging/client.rb [hunk 7]
```diff
@@ -161,18 +161,20 @@ def subscribe_background_job(options)
       end
 
       # Publish a message as a topic. All subscribers will receive a copy of the message.
-      # Expected keys in +options+ are:
+      # +messages+ can be either a hash or an array of hashes.
```
Comment (kbrock) [inline lib/manageiq/messaging/client.rb:164]: Is it possible to encourage all messages to have the same general options?
so all messages have the same target and the same sender. Possibly the same headers, too?

Or will having multiple hashes trim down on having to `map` the incoming array and to multiple hashes inside this method?

File: lib/manageiq/messaging/kafka/topic.rb [hunk 8]
```diff
@@ -8,8 +8,9 @@ module Topic
 
         private
 
-        def publish_topic_impl(options)
-          raw_publish(true, *topic_for_publish(options))
+        def publish_topic_impl(messages)
+          handles = messages.collect { |message| raw_publish(*topic_for_publish(message)) }
+          handles.each(&:wait)
```
Comment (kbrock) [inline lib/manageiq/messaging/kafka/topic.rb:13]: I like the sending and then the waiting - think this is done very nicely.

Also, nicely mimics the `queue` implementation.

File: lib/manageiq/messaging/kafka/common.rb (hunk 4 above)
Comment (kbrock) [reply lib/manageiq/messaging/kafka/common.rb:28]: 🔥

**Review by kbrock — COMMENTED** [review]
This is nice.

main comment is I would like it more if the multiple publish had a single destination. But this may be something arbitrary in my mind and not based upon use cases nor real life implementation constraints

**Review by Fryguy — APPROVED** [review]

File: lib/manageiq/messaging/kafka/common.rb (hunk 3 above)
Comment (agrare) [reply lib/manageiq/messaging/kafka/common.rb:24]: > EDIT: if raw_publish is an internal method and you moved the wait into the caller, then this is very nice.

Yes this is an internal method only

File: lib/manageiq/messaging/client.rb (hunk 7 above)
Comment (agrare) [reply lib/manageiq/messaging/client.rb:164]: I like this idea a lot, in practice I think the options will all be the same and you'll just have an array of messages.

One of the nice parts of this approach though is that code written against the current interface will still work and MIQ isn't the only consumer of this API so this has the benefits of only enhancing and not breaking the current public API.

If we want to look at adding a different method or breaking the interface, splitting the options and payload would be really interesting
